File: Day #23/solution.js
// Shortest path on a grid with walls using BFS (4-directional).
// Time: O(M*N), Space: O(M*N).
function shortestPath(board, start, end) {
  const m = board.length, n = board[0].length;
  const visited = Array.from({ length: m }, () => new Array(n).fill(false));
  const dirs = [[-1, 0], [1, 0], [0, -1], [0, 1]];
  const q = [[start[0], start[1], 0]];
  visited[start[0]][start[1]] = true;
  let head = 0;
  while (head < q.length) {
    const [r, c, d] = q[head++];
    if (r === end[0] && c === end[1]) return d;
    for (const [dr, dc] of dirs) {
      const nr = r + dr, nc = c + dc;
      if (nr >= 0 && nr < m && nc >= 0 && nc < n && !visited[nr][nc] && !board[nr][nc]) {
        visited[nr][nc] = true;
        q.push([nr, nc, d + 1]);
      }
    }
  }
  return null;
}

const f = false, t = true;
const board = [
  [f, f, f, f],
  [t, t, f, t],
  [f, f, f, f],
  [f, f, f, f],
];
const res = shortestPath(board, [3, 0], [0, 0]);
console.log(res === null ? 'None' : res);

File: Day #23/solution.rs
// Shortest path on a grid with walls using BFS (4-directional).
// Time: O(M*N), Space: O(M*N).
use std::collections::VecDeque;

fn shortest_path(board: &Vec<Vec<bool>>, start: (usize, usize), end: (usize, usize)) -> Option<i32> {
    let m = board.len();
    let n = board[0].len();
    let mut visited = vec![vec![false; n]; m];
    let dirs: [(i32, i32); 4] = [(-1, 0), (1, 0), (0, -1), (0, 1)];
    let mut q: VecDeque<(usize, usize, i32)> = VecDeque::new();
    q.push_back((start.0, start.1, 0));
    visited[start.0][start.1] = true;
    while let Some((r, c, d)) = q.pop_front() {
        if (r, c) == end {
            return Some(d);
        }
        for &(dr, dc) in dirs.iter() {
            let nr = r as i32 + dr;
            let nc = c as i32 + dc;
            if nr >= 0 && nr < m as i32 && nc >= 0 && nc < n as i32 {
                let (nr, nc) = (nr as usize, nc as usize);
                if !visited[nr][nc] && !board[nr][nc] {
                    visited[nr][nc] = true;
                    q.push_back((nr, nc, d + 1));
                }
            }
        }
    }
    None
}

fn main() {
    let f = false;
    let t = true;
    let board = vec![
        vec![f, f, f, f],
        vec![t, t, f, t],
        vec![f, f, f, f],
        vec![f, f, f, f],
    ];
    match shortest_path(&board, (3, 0), (0, 0)) {
        Some(res) => println!("{}", res),
        None => println!("None"),
    }
}
